### PR TITLE
Extension/mailers i18n

### PR DESCRIPTION
**Pull request by siddharth-vinsol**
* Send an email to the user when he places an order with all of the order info like line_item, and its image, price etc..
if the product has more than one image, then first image will be listed with line items. And other images should be attached as attachments to the email.

* Add a field in users table which would have user’s language preference: english, hindi. Default to English.
   * Then update the Order mailer to send email in user’s preferred language.
* Write a rake task, when run, should send all of the users a consolidated email of all of his orders and items.
* Change the language of error messages on all of the forms in the application. Also change all the flash messages to the user’s preferred language.
* Send the process id in email headers as X-SYSTEM-PROCESS-ID

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -31,7 +31,7 @@ def create
 
         ChargeOrderJob.perform_later(@order, pay_type_params.to_h)
 
-        format.html { redirect_to store_index_url, notice: "Thank you for your order." }
+        format.html { redirect_to store_index_url, notice: t('flash.notice.thank_you_for_order') }
```
Comment (shubhamjuneja1) [inline app/controllers/orders_controller.rb:34]: i18

File: app/controllers/products_controller.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ def create
 
     respond_to do |format|
       if @product.save
-        format.html { redirect_to product_url(@product), notice: "Product was successfully created." }
+        format.html { redirect_to product_url(@product), notice: "Product #{t('flash.notice.successfull_create')}" }
```
Comment (shubhamjuneja1) [inline app/controllers/products_controller.rb:36]: i18

File: lib/tasks/order/send_order_details.rake [hunk 3]
```diff
@@ -0,0 +1,10 @@
+namespace :order do
+  desc 'Sends all of the users a consolidated email of all his/her orders and items'
+  task :send_order_details => :environment do
+    Order.all.each do |order|
```
Comment (shubhamjuneja1) [inline lib/tasks/order/send_order_details.rake:4]: well question was to send a consolidated email to each user for all his orders
So instead it should be using a usermailer in user namespace

File: db/migrate/20220829104227_add_language_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddLanguageToUsers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :users, :language, :string, default: 'en'
```
Comment (shubhamjuneja1) [inline db/migrate/20220829104227_add_language_to_users.rb:3]: rename to language_preference

File: app/mailers/order_mailer.rb [hunk 5]
```diff
@@ -4,8 +4,19 @@ class OrderMailer < ApplicationMailer
...
   def received(order)
     @order = order
+    order.line_items.each do |line_item|
+      product = line_item.product
+      product_images = product.images.to_a
 
-    mail to: order.email, subject: 'Pragmatic Store Order Confirmation'
+      attachments.inline["#{product.title}_1"] = product_images.shift.download
+      product_images.each do |image|
+        attachments["#{product.title}_#{image.filename}"] = image.download
```
Comment (shubhamjuneja1) [inline app/mailers/order_mailer.rb:13]: Shouldn't these be inline attachments?
do we really need to download the image, can't we give the path of the image?

Comment (shubhamjuneja1) [reply app/mailers/order_mailer.rb:13]: Extract this in separate method as well

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -36,4 +37,8 @@ class ApplicationController < ActionController::Base
   protected def attach_ip_in_header
     @client_ip = request.ip
   end
+
+  protected def setup_locale(&action)
+    I18n.with_locale(User.languages[@logged_in_user.try(:language) || :en], &action)
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb:42]: you should be saving 'en' in language preference why saving as  English?
it should be something like, I18n.with_locale(current_user.preference)

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -8,10 +8,11 @@ class ApplicationController < ActionController::Base
   before_action :update_hit_counter, only: [:index, :show, :edit, :new]
   before_action :authorize
   before_action :attach_ip_in_header
+  around_action :setup_locale
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb:11]: why around_action not before_action?

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (siddharth-vinsol) [reply app/controllers/application_controller.rb:11]: discuss
https://guides.rubyonrails.org/i18n.html#managing-the-locale-across-requests

File: app/mailers/order_mailer.rb (hunk 5 above)
Comment (siddharth-vinsol) [reply app/mailers/order_mailer.rb:13]: No need todownload

**Pull request merged**